src/charm.py

```diff
@@ -0,0 +1,208 @@
+    """
+
+    _state = framework.StoredState()
+
+    @property
+    def external_hostname(self):
+        """Return the DNS listing used for external connections."""
+        return self.config["external-hostname"] or self.app.name
+
+    def __init__(self, *args):
+        """Construct.
+
+        Args:
+            args: Ignore.
+        """
+        super().__init__(*args)
+        self.name = "temporal-ui"
+
+        # Handle basic charm lifecycle.
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.temporal_ui_pebble_ready, self._on_temporal_ui_pebble_ready)
+
+        # Handle ui:temporal relation.
+        self._state.set_default(server_status="blocked")
+        self.framework.observe(self.on.ui_relation_joined, self._on_ui_relation_joined)
+        self.framework.observe(self.on.ui_relation_changed, self._on_ui_relation_changed)
+        self.framework.observe(self.on.ui_relation_broken, self._on_ui_relation_broken)
+
+        # Handle Ingress.
+        self.ingress = IngressRequires(
+            self,
+            {
+                "service-hostname": self.external_hostname,
+                "service-name": self.app.name,
+                "service-port": 8080,
+            },
+        )
+
+    @log_event_handler(logger)
+    def _on_install(self, event):
+        """Install temporal UI tools.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.unit.status = MaintenanceStatus("installing temporal ui tools")
+
+    @log_event_handler(logger)
+    def _on_temporal_ui_pebble_ready(self, event):
+        """Define and start temporal UI using the Pebble API.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self._update(event)
+
+    @log_event_handler(logger)
+    def _on_config_changed(self, event):
+        """Handle configuration changes.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.unit.status = WaitingStatus("configuring temporal")
+        self.ingress.update_config({"service-hostname": self.external_hostname})
+        self._update(event)
+
+    @log_event_handler(logger)
+    def _on_ui_relation_joined(self, event):
+        """Handle joining a ui:temporal relation.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.unit.status = WaitingStatus(f"handling {event.relation.name} change")
+        self._state.server_status = event.relation.data[event.app].get("server_status")
+        self._update(event)
+
+    @log_event_handler(logger)
+    def _on_ui_relation_changed(self, event):
+        """Handle changes on the ui:temporal relation.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self._state.server_status = event.relation.data[event.app].get("server_status")
```

Comment:
So I see in this case the use of _state is fine but it seems like you're only using the stored state to store something coming out of a peer relation and then using `_state` for logging. In which case you could just hold `server_status` in a variable in the class I think.